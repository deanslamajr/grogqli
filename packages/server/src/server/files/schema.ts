import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
