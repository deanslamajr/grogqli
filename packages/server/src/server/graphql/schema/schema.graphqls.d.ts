/* 19a699f8a8761279db5b06650f17155319410b52
 * This file is automatically generated by graphql-let. */

import { GraphQLResolveInfo } from 'graphql';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
    [key: string]: unknown;
}> = {
    [K in keyof T]: T[K];
};
/** All built-in and custom scalars, mapped to their actual values */
export declare type Scalars = {
    ID: string;
    String: string;
    Boolean: boolean;
    Int: number;
    Float: number;
};
export declare type Recording = {
    __typename?: 'Recording';
    replaceThisBullshxt: Scalars['String'];
};
export declare type Query = {
    __typename?: 'Query';
    getRecordings: Array<Maybe<Recording>>;
};
export declare type ResolverTypeWrapper<T> = Promise<T> | T;
export declare type LegacyStitchingResolver<TResult, TParent, TContext, TArgs> = {
    fragment: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type NewStitchingResolver<TResult, TParent, TContext, TArgs> = {
    selectionSet: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type StitchingResolver<TResult, TParent, TContext, TArgs> = LegacyStitchingResolver<TResult, TParent, TContext, TArgs> | NewStitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | StitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type ResolverFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => Promise<TResult> | TResult;
export declare type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => AsyncIterator<TResult> | Promise<AsyncIterator<TResult>>;
export declare type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<{
        [key in TKey]: TResult;
    }, TParent, TContext, TArgs>;
    resolve?: SubscriptionResolveFn<TResult, {
        [key in TKey]: TResult;
    }, TContext, TArgs>;
}
export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
    resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}
export declare type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> = SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs> | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;
export declare type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> = ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>) | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;
export declare type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (parent: TParent, context: TContext, info: GraphQLResolveInfo) => Maybe<TTypes> | Promise<Maybe<TTypes>>;
export declare type IsTypeOfResolverFn<T = {}> = (obj: T, info: GraphQLResolveInfo) => boolean | Promise<boolean>;
export declare type NextResolverFn<T> = () => Promise<T>;
export declare type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (next: NextResolverFn<TResult>, parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
/** Mapping between all available schema types and the resolvers types */
export declare type ResolversTypes = {
    Recording: ResolverTypeWrapper<Recording>;
    String: ResolverTypeWrapper<Scalars['String']>;
    Query: ResolverTypeWrapper<{}>;
    Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
};
/** Mapping between all available schema types and the resolvers parents */
export declare type ResolversParentTypes = {
    Recording: Recording;
    String: Scalars['String'];
    Query: {};
    Boolean: Scalars['Boolean'];
};
export declare type RecordingResolvers<ContextType = any, ParentType extends ResolversParentTypes['Recording'] = ResolversParentTypes['Recording']> = {
    replaceThisBullshxt?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
    __isTypeOf?: IsTypeOfResolverFn<ParentType>;
};
export declare type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
    getRecordings?: Resolver<Array<Maybe<ResolversTypes['Recording']>>, ParentType, ContextType>;
};
export declare type Resolvers<ContextType = any> = {
    Recording?: RecordingResolvers<ContextType>;
    Query?: QueryResolvers<ContextType>;
};
/**
 * @deprecated
 * Use "Resolvers" root object instead. If you wish to get "IResolvers", add "typesPrefix: I" to your config.
 */
export declare type IResolvers<ContextType = any> = Resolvers<ContextType>;

          
// This is an extra code in addition to what graphql-codegen makes.
// Users are likely to use 'graphql-tag/loader' with 'graphql-tag/schema/loader'
// in webpack. This code enables the result to be typed.
import { DocumentNode } from 'graphql'
export default typeof DocumentNode
