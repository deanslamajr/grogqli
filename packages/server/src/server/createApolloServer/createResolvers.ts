import { IObjectTypeResolver } from 'graphql-tools';
import { Config } from 'apollo-server-core';

import { SchemaFile } from '../files';
import { openTypeNameToIdMapping } from '../files/type';
import { loadOperationsMappingFile } from '../files/operation';
import { fieldResolverFactory } from './fieldResolverFactory';

type Resolvers = NonNullable<Config['resolvers']>;

// TODO add logic to check a new "version" field on the schema and version the createResolver algorithm
// perhaps put the versioning in the filename so that the structure of the file doesn't need modification
// This might be more ideal as the original (current) version (v1) maps directly to the IntrospectionQuery type
export const createResolvers = async ({
  introspectionQuery: schema,
  id: schemaId,
}: SchemaFile): Promise<Resolvers> => {
  const [operationsData, typeNameToIdMappingData] = await Promise.all([
    loadOperationsMappingFile(schemaId),
    openTypeNameToIdMapping(schemaId),
  ]);

  if (operationsData === null) {
    // TODO handle case where file doesnt exist for the given schemaId
    throw new Error(
      `TODO handle case where a operations file doesnt exist for the given schemaId. schemaId:${schemaId}`
    );
  }

  return schema.__schema.types.reduce<Resolvers>((resolvers, CurrentType) => {
    // add a resolver for each object type in the schema
    // note: skip graphql internal types (those prefixed with '__')
    if (CurrentType.kind === 'OBJECT' && !CurrentType.name.includes('__')) {
      const currentTypeName = CurrentType.name;

      const resolverForCurrentType = CurrentType.fields.reduce<
        IObjectTypeResolver
      >((resolver, { name: fieldName }) => {
        const resolveField = fieldResolverFactory({
          operationsData,
          typeNameToIdMappingData,
          schema,
          parentTypeName: currentTypeName,
          fieldName,
        });
        resolver[fieldName] = resolveField;
        return resolver;
      }, {} as IObjectTypeResolver);

      resolvers[currentTypeName] = resolverForCurrentType;
    }
    return resolvers;
  }, {} as Resolvers);
};
