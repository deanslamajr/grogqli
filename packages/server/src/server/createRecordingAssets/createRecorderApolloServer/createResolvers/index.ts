import { IObjectTypeResolver } from 'graphql-tools';
import { Config } from 'apollo-server-core';

import { SchemaFile } from '../../../files';
import { fieldResolverFactory } from './fieldResolverFactory';

type Resolvers = NonNullable<Config['resolvers']>;

export const createResolvers = async ({
  introspectionQuery: schema,
}: SchemaFile): Promise<Resolvers> => {
  return schema.__schema.types
    .filter(({ name }) => !name.includes('__'))
    .reduce<Resolvers>((resolvers, CurrentType) => {
      // add a resolver for each object type in the schema
      // note: skip graphql internal types (those prefixed with '__')
      if (CurrentType.kind === 'OBJECT') {
        const currentTypeName = CurrentType.name;

        const resolverForCurrentType = CurrentType.fields.reduce<
          IObjectTypeResolver
        >((resolver, { name: fieldName, type }) => {
          const resolveField = fieldResolverFactory({
            returnType: type,
            schema,
            parentTypeName: currentTypeName,
            fieldName,
          });
          resolver[fieldName] = resolveField;
          return resolver;
        }, {} as IObjectTypeResolver);

        resolvers[currentTypeName] = resolverForCurrentType;
      }
      return resolvers;
    }, {} as Resolvers);
};
