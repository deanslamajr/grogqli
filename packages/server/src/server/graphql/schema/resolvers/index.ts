import {Resolvers} from '@grogqli/schema';

import { queryResolver } from './queries';
import {mutationResolver } from './mutations';
import {subscriptionResolver} from './subscriptions';

export const resolvers: Partial<Resolvers> = {
  Query: queryResolver,
  Mutation: mutationResolver,
  Subscription: subscriptionResolver
};
