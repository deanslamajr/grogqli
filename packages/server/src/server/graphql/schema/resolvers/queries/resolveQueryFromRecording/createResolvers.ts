import { IResolvers, IObjectTypeResolver } from 'graphql-tools';

import { getOperationsData, OperationsData, SchemaFile } from '../../files';
import { fieldResolverFactory } from './fieldResolverFactory';

// TODO add logic to check a new "version" field on the schema and version the createResolver algorithm
// perhaps put the versioning in the filename so that the structure of the file doesn't need modification
// This might be more ideal as the original (current) version (v1) maps directly to the IntrospectionQuery type
export const createResolvers = async ({
  introspectionQuery: schema,
  id: schemaId,
}: SchemaFile): Promise<IResolvers> => {
  const operationsData: OperationsData = await getOperationsData();
  return schema.__schema.types.reduce<IResolvers>((resolvers, CurrentType) => {
    // add a resolver for each object type in the schema
    // note: skip graphql internal types (those prefixed with '__')
    if (CurrentType.kind === 'OBJECT' && !CurrentType.name.includes('__')) {
      const currentTypeName = CurrentType.name;

      const resolverForCurrentType = CurrentType.fields.reduce<
        IObjectTypeResolver
      >((resolver, { name: fieldName }) => {
        const resolveField = fieldResolverFactory({
          operationsData,
          schema,
          parentTypeName: currentTypeName,
          fieldName,
        });
        resolver[fieldName] = resolveField;
        return resolver;
      }, {} as IObjectTypeResolver);

      resolvers[currentTypeName] = resolverForCurrentType;
    }
    return resolvers;
  }, {} as IResolvers);
};
