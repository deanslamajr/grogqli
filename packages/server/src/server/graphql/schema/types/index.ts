import { join } from 'path'
import { loadFilesSync } from '@graphql-tools/load-files'
import { mergeTypeDefs } from '@graphql-tools/merge'

const loadedFiles = loadFilesSync(join(process.cwd(), '**/*.graphqls'))
export const typeDefs = mergeTypeDefs(loadedFiles)