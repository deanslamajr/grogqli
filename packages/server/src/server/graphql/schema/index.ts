import { makeExecutableSchema } from '@graphql-tools/schema';
import { typeDefs } from '@grogqli/schema';

import { resolvers } from './resolvers';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
