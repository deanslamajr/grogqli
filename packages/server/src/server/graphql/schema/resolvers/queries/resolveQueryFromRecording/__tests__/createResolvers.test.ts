import { createResolvers } from '../createResolvers';
import schema from './schemas/schema.json';

describe('createResolvers', () => {
  xit('should return a resolver configuration that includes each type in the schema (ignoring gql internal types)', () => {
    const actual = createResolvers(schema);
    expect(actual).toMatchSnapshot();
  });
});
