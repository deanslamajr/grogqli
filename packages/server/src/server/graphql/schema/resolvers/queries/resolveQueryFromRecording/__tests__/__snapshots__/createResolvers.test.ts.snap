// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResolvers should return a resolver configuration that includes each type in the schema (ignoring gql internal types) 1`] = `
Object {
  "Account": Object {
    "email": undefined,
  },
  "Chore": Object {
    "description": undefined,
    "dueDate": undefined,
    "id": undefined,
    "summary": undefined,
    "version": undefined,
  },
  "CompleteChoreResponse": Object {
    "hasAccountSession": undefined,
    "wasSuccess": undefined,
  },
  "CreateChoreResponse": Object {
    "hasAccountSession": undefined,
    "newChore": undefined,
    "wasCreateSuccess": undefined,
  },
  "FinishAccountCreateResponse": Object {
    "accountCreateSuccess": undefined,
    "wasTokenValid": undefined,
  },
  "GetChoresResponse": Object {
    "chores": undefined,
    "hasAccountSession": undefined,
  },
  "LoginAccountResponse": Object {
    "wasLoginSuccess": undefined,
  },
  "LogoutAccountResponse": Object {
    "wasLogoutSuccess": undefined,
  },
  "Mutation": Object {
    "completeChore": undefined,
    "createChore": undefined,
    "finishAccountCreate": undefined,
    "loginAccount": undefined,
    "logoutAccount": undefined,
    "sendAccountCreateEmail": undefined,
    "updateChore": undefined,
  },
  "Query": Object {
    "getAccountFromSession": undefined,
    "getChores": undefined,
  },
  "SendAccountCreateEmailResponse": Object {
    "emailSendSuccess": undefined,
    "recaptchaPassed": undefined,
  },
  "UpdateChoreResponse": Object {
    "doesChoreExist": undefined,
    "hasAccountSession": undefined,
    "updatedChore": undefined,
    "wasUpdateSuccess": undefined,
  },
}
`;
